Changing the frontend to Tailwind CSS and more

Dominik1999 opened the pull request:
* Frontend changed to Tailwind CSS (mui only needed for dropdown), closes #7 
* Added table dummy for Miden assembly instructions, starts #43 
* Added more buttons (debug, prove, verify)
* Output looks better, closes #45
* Added Miden Logo, closes #16

Dominik1999 commented on src/App.tsx:
```diff
@@ -1,21 +1,14 @@
 import {
```
Unfortunately, I still need some MUI elements for the dropdown. I wasn't able to use a Tailwind CSS dropdown that also vanishes when I click somewhere. Also there I have to import a library that uses Tailwind. But maybe one of you have an idea?

grjte replied:
We should get rid of MUI. I have ideas for this - let's discuss separately

Dominik1999 replied:
No more MUI.

Dominik1999 commented on src/App.tsx:
```diff
@@ -78,107 +67,268 @@ end`
-        </Box>
+        <div className="flex items-center justify-between flex-wrap bg-blue-700 p-6">
+          <div className="flex items-center flex-shrink-0 text-white mr-6">
+            <svg className="fill-white" width="100.333" height="100.333" version="1.0" viewBox="0 0 400 400"><path d="M187.8 42.5c-4.7 1.3-7.7 2.9-64.8 35.9-50.9 29.5-50.4 29.2-55.7 35-5.5 5.9-7.9 10.4-9.9 18.1-2 7.4-2 129.7 0 137.1 3 11.2 11.2 21.6 21.8 27.6 4.5 2.5 34.4 19.8 74.3 42.9 32.4 18.7 35.1 19.9 45.4 19.9 8.9 0 16.6-2.7 30-10.4 6.9-4 20.3-11.7 29.6-17.1 9.4-5.3 21.7-12.5 27.5-16 5.8-3. ...
```
This is only the Miden Logo

Dominik1999 replied:
Jup, I had this before. Would a separate file be better? Isn't the website faster to load when using a picture in this style (might not be super relevant here though).

Dominik1999 replied:
perfect, done

grjte reviewed (changes requested):
Looks good! I like the way the inputs and outputs are now. I think beyond that we'll want to get a design anyway, so it works for now.

Regarding the implementation - I left a few notes inline. The main comment I have is that there's a lot of duplication here, and each duplicated container should be pulled out into a component instead that accepts a set of props. I gave an example with some of the buttons, but this is also true of anything that's repeated with the same styling (e.g. the headers for "inputs"/"outputs"/"Miden Assembly Code", the rows/cells of the grid, etc). We can come back through and do a bigger refactor later, but it would be great to do some of it now.

grjte commented on src/App.tsx:
```diff
@@ -78,107 +67,268 @@ end`
+            </FormControl>
 
-      </Container>
+            <button className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded" onClick={() => {
+            init().then(() => {
+              try {
+                let resp = program(code, inputs, numOfOutputs);
+                console.log(resp.toString());
+                setOutput(`Miden VM Program Output \nStack = [${resp.toString()}] \nCycles = <need to return cycles> \n`);
+              } catch (error) {
+                setOutput("Error: Check the developer console for details.");
+              }
+                });
+              }}>
+            Run
+            </button>
```
You can look up presentational vs. container components in React to learn more about this design pattern

Dominik1999 replied:
nice, that makes more sense and it makes everything cleaner

Dominik1999 commented on src/App.tsx:
```diff
@@ -78,107 +67,268 @@ end`
+            Run
+            </button>
+
+            <button className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
+            Debug
+            </button>
+            <button className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
+            Prove
+            </button>
+            <button className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
+            Verify
+            </button>
```
Ah got it, yes that is way cleaner. Let me try to apply this pattern on more components

grjte commented on src/App.tsx:
```diff
@@ -60,10 +57,16 @@ end`
   const [example, setExample] = React.useState<string>();
-  const handleSelectChange = async (event: SelectChangeEvent) => {
-    const value = event.target.value;
+  console.log(example)
+  const handleSelectChange = async (exampleChange: string) => {
+    const value = exampleChange;
+    console.log("adad")
```
Should we remove these console logs?

grjte commented on src/App.tsx:
```diff
@@ -76,109 +79,96 @@ end`
+          </div>
+          <div className="w-full block flex-grow lg:flex lg:items-center lg:w-auto">
+            <div className="text-sm lg:flex-grow">
+              <a href="https://wiki.polygon.technology/docs/miden/user_docs/assembly/main" className="block mt-4 lg:inline-block lg:mt-0 text-teal-200 hover:text-white mr-4">
+                Docs
+              </a>
+              <a href="https://github.com/0xPolygonMiden/examples#available-examples" className="block mt-4 lg:inline-block lg:mt-0 text-teal-200 hover:text-white mr-4" target="_blank" rel="noopener noreferrer">
+                README: Examples
+              </a>
+              <a href="https://polygon.technology/solutions/polygon-miden/" className="block mt-4 lg:inline-block lg:mt-0 text-teal-200 hover:text-white">
+                Homepage
+              </a>
```
I would probably create a `Link` component or at least pull the custom styling out into a `globals.css` file so it's not duplicated. https://tailwindcss.com/docs/adding-custom-styles#adding-component-classes

grjte commented on src/components/ActionButton.tsx:
```diff
@@ -0,0 +1,18 @@
+type ActionButtonProps = {
+    label: string,
+    onClick?: any,
```
nit: ideally, we should avoid using "any" types. You should be able to use the following:

```
onClick?: () => void;
```

grjte commented on src/components/DropDown.tsx:
```diff
@@ -0,0 +1,88 @@
+
+export default function DropDown({ onExampleValueChange }: DropDownProps) {
+  const [selected, setSelected] = useState(examples[1])
+  //console.log(selected.name)
```
We can probably remove this

grjte commented on src/components/DropDown.tsx:
```diff
@@ -0,0 +1,88 @@
+    </Listbox>
+  )
+}
+function setTags(arg0: string[]) {
+  throw new Error('Function not implemented.')
+}
```
do we need this?

Dominik1999 replied:
nope

grjte commented on package.json:
```diff
@@ -5,14 +5,17 @@
   "dependencies": {
     "@codemirror/theme-one-dark": "^6.1.0",
+    "@headlessui/react": "^1.7.7",
+    "@heroicons/react": "^2.0.13",
```
Since we're not using very many icons, we could just copy the code for the few that we need from the site and remove this dependency. https://heroicons.com/

Dominik1999 replied:
you are right

grjte commented on src/components/InstructionTable.tsx:
```diff
@@ -0,0 +1,96 @@
+                    <th scope="col" className="py-3.5 pl-4 pr-3 text-left text-sm font-semibold text-gray-900 sm:pl-6">
+                      Instruction
+                    </th>
+                    <th scope="col" className="px-3 py-3.5 text-left text-sm font-semibold text-gray-900">
+                      Stack Input
+                    </th>
+                    <th scope="col" className="px-3 py-3.5 text-left text-sm font-semibold text-gray-900">
+                      Stack Output
+                    </th>
+                    <th scope="col" className="px-3 py-3.5 text-left text-sm font-semibold text-gray-900">
+                      Cycles
+                    </th>
+                    <th scope="col" className="px-3 py-3.5 text-left text-sm font-semibold text-gray-900">
```
I would define a named for "px-3 py-3.5 text-left text-sm font-semibold text-gray-900" in `globals.css` so it doesn't need to be duplicated

Dominik1999 replied:
done

grjte commented on src/App.tsx:
```diff
@@ -1,57 +1,54 @@
 
-  // this should be dyncamically read from ./examples/ - buy my TS knowledge is limited
-  const examples = [
+  /**
+  * This lists all the examples available in the examples folder.
+  */  
+ const examples = [
```
Do we still need this defined here? It's also defined in the `Dropdown` component.

Dominik1999 replied:
nope its a left over

grjte commented on src/App.tsx:
```diff
@@ -1,57 +1,54 @@
   }
 
-function App() {
+function App(this: any) {
```
why do we need this `this: any`?

grjte reviewed (changes requested):
Thanks @Dominik1999 this refactor looks great! I left a few small comments inline. Aside from those, the final thing I'd recommend changing is to reduce the duplication in the styles. We can define custom styles with tailwind so components/tags that share the same 5 style items can each get one shared style name instead. I left a couple of related comments inline, and you can learn more here: https://tailwindcss.com/docs/adding-custom-styles#adding-component-classes

Feel free to ping me directly as well

grjte reviewed (changes requested):
Looks great now! I just noticed that there's still a dependency for the `@mui/icons-material`. Do we still need this? Also, I would have expected the `package-lock.json` to change after you removed the dependency on the heroicons, but it didn't.

Can you just double check on the dependencies to make sure we've cleaned everything up? After that I think we're ready to merge

Dominik1999 commented:
> Looks great now! I just noticed that there's still a dependency for the `@mui/icons-material`. Do we still need this? Also, I would have expected the `package-lock.json` to change after you removed the dependency on the heroicons, but it didn't.
> 
> Can you just double check on the dependencies to make sure we've cleaned everything up? After that I think we're ready to merge

Perfect, I removed the dependent and another file. I think we can merge.

grjte reviewed (approved):
Yep, good catch with the `theme.ts` file as well. Looks good, thanks!
